fs/readdircount_linux_amd64.go

```diff
@@ -0,0 +1,67 @@
+// +build linux,amd64
+
+package fs
+
+import (
+	"errors"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+func countDirEntries(buf []byte, n int) int {
+	count := 0
+	buf = buf[:n]
+	for len(buf) > 0 {
+		// see man page getdents(2) for struct linux_dirent64
+		reclen := *(*uint16)(unsafe.Pointer(&buf[16]))
+		ino := *(*uint64)(unsafe.Pointer(&buf[0]))
+
+		if int(reclen) > len(buf) {
+			return count
+		}
+		buf = buf[reclen:]
+		if ino == 0 {
```

Comment:
OK, please add a comment.

> I don't think that procfs returns such entries

It seems it does, so please adjust the code accordingly.

```
$ ls -la /proc/self/fd/
total 0
dr-x------ 2 ubuntu ubuntu  0 Apr 25 14:18 .
dr-xr-xr-x 9 ubuntu ubuntu  0 Apr 25 14:18 ..
lrwx------ 1 ubuntu ubuntu 64 Apr 25 14:18 0 -> /dev/pts/1
lrwx------ 1 ubuntu ubuntu 64 Apr 25 14:18 1 -> /dev/pts/1
lrwx------ 1 ubuntu ubuntu 64 Apr 25 14:18 2 -> /dev/pts/1
lr-x------ 1 ubuntu ubuntu 64 Apr 25 14:18 3 -> /proc/12025/fd
```

How did you test this? I am surprised you didn't notice the off-by-2 in the results.